File: 0-environment-setup-and-nestjs-core/src/cat/cat.controller.ts
/**
 * Controller xử lý endpoint `/cats` — demo Module + DI.
 * (EN: Controller handling `/cats` endpoint — Module + DI demo.)
 */
import {
    Controller,
    Get,
} from "@nestjs/common"
import {
    CatService,
} from "./cat.service"

@Controller("cats")
export class CatController {
    /**
     * Inject `CatService` để controller chỉ giữ routing responsibility.
     * (EN: Inject `CatService` so controller only keeps routing responsibility.)
     */
    constructor(private readonly catService: CatService) {}

    /**
     * Xử lý `GET /cats` và trả dữ liệu từ service.
     * (EN: Handle `GET /cats` and return data from service.)
     */
    @Get()
    findAll() {
        // Delegate qua service để controller không chứa business/data logic.
        // (EN: Delegate to service to keep controller free of business/data logic.)
        return this.catService.getCats()
    }
}

File: 0-frameworks-in-backend/backend/1-java/src/main/java/com/example/cat/CatController.java
package com.example.cat;

import java.util.List;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * REST controller for the {@code /cats} endpoint.
 *
 * <p>{@code @RestController} combines {@code @Controller} + {@code @ResponseBody}.
 * Spring resolves the {@link CatService} dependency via constructor injection — no
 * manual {@code new} required.
 */
@RestController
public class CatController {

    /** Injected by Spring's IoC container; controller stays free of data logic. */
    private final CatService catService;

    /**
     * Constructor injection — Spring detects the single constructor and supplies
     * the matching bean from its context.
     *
     * @param catService the singleton {@link CatService} bean managed by Spring
     */
    public CatController(CatService catService) {
        this.catService = catService;
    }

    /**
     * Handle {@code GET /cats} — demonstrates controller → service delegation.
     *
     * @return list of all cats; Jackson serializes each as {@code {"id":.,"name":..}}
     */
    @GetMapping("/cats")
    public List<Cat> cats() {
        // Delegate to service; controller only routes the request.
        return catService.all();
    }
}
